gstreamer-source.c

```diff
@@ -67,11 +77,53 @@ static gboolean start_pipe(gpointer user_data)
 	return G_SOURCE_REMOVE;
 }
 
+static void update_obs_media_state(GstMessage *message, data_t *data)
+{
+	switch (GST_MESSAGE_TYPE(message)) {
+	case GST_MESSAGE_BUFFERING: {
+		gint percent;
+		gst_message_parse_buffering(message, &percent);
+		data->buffering = (percent < 100);
+	} break;
+	case GST_MESSAGE_STATE_CHANGED: {
+		GstState newstate;
+		gst_message_parse_state_changed(message, NULL, &newstate, NULL);
+		switch (newstate) {
+		default:
+		case GST_STATE_NULL:
+			data->obs_media_state = OBS_MEDIA_STATE_NONE;
+			break;
+		case GST_STATE_READY:
+			data->obs_media_state = OBS_MEDIA_STATE_STOPPED;
+			break;
+		case GST_STATE_PAUSED:
+			data->obs_media_state = data->buffering
+					    ? OBS_MEDIA_STATE_BUFFERING
+					    : OBS_MEDIA_STATE_PAUSED;
+			break;
+		case GST_STATE_PLAYING:
```

Comment:
I think you need to make the same buffering logic as for PAUSED. A pipeline does not change its state just because it is buffering.

If it is desired to pause and wait, the application or code would have to do so. But that would increase latency for that stream. That is something for another day though.

Reply:
Or actually, maybe it would be more correct to just save the state here and when the actual sate is requested, first check for the `buffering` flag, return `OBS_MEDIA_STATE_BUFFERING` if it is set, and only return the real state if the flag is cleared.

Reply:
You're right on both counts.  I was not doing the "buffering override" for PLAYING because it seemed to cause bad behavior which I now realize was actually because of your second comment...  It is actually required that we evaluate our buffering override at query-time because we might not get any further state change messages after receiving the "buffering=100%" message.

Addressed in my [v4] push.